diff -u -r a/ymfm.h b/ymfm.h
--- a/ymfm.h
+++ b/ymfm.h
@@ -33,6 +33,8 @@
 
 #pragma once
 
+#include "ymfm_owcompat.h"
+
 #if defined(_MSC_VER) && !defined(_CRT_SECURE_NO_WARNINGS)
  #define _CRT_SECURE_NO_WARNINGS
 #endif
@@ -69,12 +71,12 @@
 	static constexpr bool LOG_UNEXPECTED_READ_WRITES = false;
 
 	// helpers to write based on the log type
-	template<typename... Params> static void log_fm_write(Params &&... args) { if (LOG_FM_WRITES) log(args...); }
-	template<typename... Params> static void log_keyon(Params &&... args) { if (LOG_KEYON_EVENTS) log(args...); }
-	template<typename... Params> static void log_unexpected_read_write(Params &&... args) { if  (LOG_UNEXPECTED_READ_WRITES) log(args...); }
+	static void log_fm_write(const char *, ...) { }
+	static void log_keyon(const char *, ...) { }
+	static void log_unexpected_read_write(const char *, ...) { }
 
 	// downstream helper to output log data; defaults to printf
-	template<typename... Params> static void log(Params &&... args) { printf(args...); }
+	static void log(const char *, ...) { }
 };
 
 
@@ -255,7 +257,7 @@
 //*********************************************************
 
 // various envelope states
-enum envelope_state : uint32_t
+enum envelope_state
 {
 	EG_DEPRESS = 0,		// OPLL only; set EG_HAS_DEPRESS to enable
 	EG_ATTACK = 1,
@@ -267,7 +269,7 @@
 };
 
 // external I/O access classes
-enum access_class : uint32_t
+enum access_class
 {
 	ACCESS_IO = 0,
 	ACCESS_ADPCM_A,
@@ -420,10 +422,7 @@
 	template<typename DataType>
 	void save_restore(DataType &data)
 	{
-		if (saving())
-			save(data);
-		else
-			restore(data);
+		(void)data;
 	}
 
 public:
@@ -474,7 +473,7 @@
 class ymfm_engine_callbacks
 {
 public:
-	virtual ~ymfm_engine_callbacks() = default;
+	virtual ~ymfm_engine_callbacks() { }
 
 	// timer callback; called by the interface when a timer fires
 	virtual void engine_timer_expired(uint32_t tnum) = 0;
@@ -494,10 +493,9 @@
 class ymfm_interface
 {
 	// the engine is our friend
-	template<typename RegisterType> friend class fm_engine_base;
 
 public:
-	virtual ~ymfm_interface() = default;
+	virtual ~ymfm_interface() { }
 
 	// the following functions must be implemented by any derived classes; the
 	// default implementations are sufficient for some minimal operation, but will
@@ -555,7 +553,7 @@
 	// of the chip; our responsibility is to pass the written data on to any consumers
 	virtual void ymfm_external_write(access_class type, uint32_t address, uint8_t data) { }
 
-protected:
+public:
 	// pointer to engine callbacks -- this is set directly by the engine at
 	// construction time
 	ymfm_engine_callbacks *m_engine;
diff -u -r a/ymfm_fm.h b/ymfm_fm.h
--- a/ymfm_fm.h
+++ b/ymfm_fm.h
@@ -43,7 +43,7 @@
 //*********************************************************
 
 // three different keyon sources; actual keyon is an OR over all of these
-enum keyon_type : uint32_t
+enum keyon_type
 {
 	KEYON_NORMAL = 0,
 	KEYON_RHYTHM = 1,
@@ -74,7 +74,8 @@
 	uint32_t multiple;                // multiple value (x.1, used to compute phase_step)
 	uint32_t eg_sustain;              // sustain level, shifted up to envelope values
 	uint8_t eg_rate[EG_STATES];       // envelope rate, including KSR
-	uint8_t eg_shift = 0;             // envelope shift amount
+	uint8_t eg_shift;                 // envelope shift amount
+	opdata_cache() : eg_shift(0) { }
 };
 
 
@@ -249,7 +250,7 @@
 template<class RegisterType>
 class fm_channel
 {
-	using output_data = ymfm_output<RegisterType::OUTPUTS>;
+	typedef ymfm_output<RegisterType::OUTPUTS> output_data;
 
 public:
 	// constructor
@@ -267,7 +268,7 @@
 	// assign operators
 	void assign(uint32_t index, fm_operator<RegisterType> *op)
 	{
-		assert(index < m_op.size());
+		assert(index < 4);
 		m_op[index] = op;
 		if (op != nullptr)
 			op->set_choffs(m_choffs);
@@ -330,7 +331,7 @@
 	uint32_t m_choffs;                     // channel offset in registers
 	int16_t m_feedback[2];                 // feedback memory for operator 1
 	mutable int16_t m_feedback_in;         // next input value for op 1 feedback (set in output)
-	std::array<fm_operator<RegisterType> *, 4> m_op; // up to 4 operators
+	fm_operator<RegisterType> *m_op[4]; // up to 4 operators
 	RegisterType &m_regs;                  // direct reference to registers
 	fm_engine_base<RegisterType> &m_owner; // reference to the owning engine
 };
@@ -358,7 +359,7 @@
 	static constexpr uint8_t STATUS_IRQ = RegisterType::STATUS_IRQ;
 
 	// expose the correct output class
-	using output_data = ymfm_output<OUTPUTS>;
+	typedef ymfm_output<OUTPUTS> output_data;
 
 	// constructor
 	fm_engine_base(ymfm_interface &intf);
@@ -418,8 +419,8 @@
 	void invalidate_caches() { m_modified_channels = RegisterType::ALL_CHANNELS; }
 
 	// simple getters for debugging
-	fm_channel<RegisterType> *debug_channel(uint32_t index) const { return m_channel[index].get(); }
-	fm_operator<RegisterType> *debug_operator(uint32_t index) const { return m_operator[index].get(); }
+	fm_channel<RegisterType> *debug_channel(uint32_t index) const { return m_channel[index]; }
+	fm_operator<RegisterType> *debug_operator(uint32_t index) const { return m_operator[index]; }
 
 public:
 	// timer callback; called by the interface when a timer fires
@@ -451,8 +452,8 @@
 	uint32_t m_modified_channels;    // mask of channels that have been modified
 	uint32_t m_prepare_count;        // counter to do periodic prepare sweeps
 	RegisterType m_regs;             // register accessor
-	std::unique_ptr<fm_channel<RegisterType>> m_channel[CHANNELS]; // channel pointers
-	std::unique_ptr<fm_operator<RegisterType>> m_operator[OPERATORS]; // operator pointers
+	fm_channel<RegisterType> *m_channel[CHANNELS]; // channel pointers
+	fm_operator<RegisterType> *m_operator[OPERATORS]; // operator pointers
 #if (YMFM_DEBUG_LOG_WAVFILES)
 	mutable ymfm_wavfile<1> m_wavfile[CHANNELS]; // for debugging
 #endif
diff -u -r a/ymfm_fm.ipp b/ymfm_fm.ipp
--- a/ymfm_fm.ipp
+++ b/ymfm_fm.ipp
@@ -797,12 +797,12 @@
 template<class RegisterType>
 fm_channel<RegisterType>::fm_channel(fm_engine_base<RegisterType> &owner, uint32_t choffs) :
 	m_choffs(choffs),
-	m_feedback{ 0, 0 },
 	m_feedback_in(0),
-	m_op{ nullptr, nullptr, nullptr, nullptr },
 	m_regs(owner.regs()),
 	m_owner(owner)
 {
+	m_feedback[0] = m_feedback[1] = 0;
+	m_op[0] = m_op[1] = m_op[2] = m_op[3] = 0;
 }
 
 
@@ -839,12 +839,12 @@
 template<class RegisterType>
 void fm_channel<RegisterType>::keyonoff(uint32_t states, keyon_type type, uint32_t chnum)
 {
-	for (uint32_t opnum = 0; opnum < m_op.size(); opnum++)
+	for (uint32_t opnum = 0; opnum < 4; opnum++)
 		if (m_op[opnum] != nullptr)
 			m_op[opnum]->keyonoff(bitfield(states, opnum), type);
 
 	if (debug::LOG_KEYON_EVENTS && ((debug::GLOBAL_FM_CHANNEL_MASK >> chnum) & 1) != 0)
-		for (uint32_t opnum = 0; opnum < m_op.size(); opnum++)
+		for (uint32_t opnum = 0; opnum < 4; opnum++)
 			if (m_op[opnum] != nullptr)
 				debug::log_keyon("%c%s\n", bitfield(states, opnum) ? '+' : '-', m_regs.log_keyon(m_choffs, m_op[opnum]->opoffs()).c_str());
 }
@@ -860,7 +860,7 @@
 	uint32_t active_mask = 0;
 
 	// prepare all operators and determine if they are active
-	for (uint32_t opnum = 0; opnum < m_op.size(); opnum++)
+	for (uint32_t opnum = 0; opnum < 4; opnum++)
 		if (m_op[opnum] != nullptr)
 			if (m_op[opnum]->prepare())
 				active_mask |= 1 << opnum;
@@ -880,7 +880,7 @@
 	m_feedback[0] = m_feedback[1];
 	m_feedback[1] = m_feedback_in;
 
-	for (uint32_t opnum = 0; opnum < m_op.size(); opnum++)
+	for (uint32_t opnum = 0; opnum < 4; opnum++)
 		if (m_op[opnum] != nullptr)
 			m_op[opnum]->clock(env_counter, lfo_raw_pm);
 
@@ -888,9 +888,9 @@
 useful temporary code for envelope debugging
 if (m_choffs == 0x101)
 {
-	for (uint32_t opnum = 0; opnum < m_op.size(); opnum++)
+	for (uint32_t opnum = 0; opnum < 4; opnum++)
 	{
-		auto &op = *m_op[((opnum & 1) << 1) | ((opnum >> 1) & 1)];
+		fm_operator<RegisterType> &op = *m_op[((opnum & 1) << 1) | ((opnum >> 1) & 1)];
 		printf(" %c%03X%c%c ",
 			"PADSRV"[op.debug_eg_state()],
 			op.debug_eg_attenuation(),
@@ -1184,22 +1184,23 @@
 	m_clock_prescale(RegisterType::DEFAULT_PRESCALE),
 	m_irq_mask(STATUS_TIMERA | STATUS_TIMERB),
 	m_irq_state(0),
-	m_timer_running{0,0},
 	m_total_clocks(0),
 	m_active_channels(ALL_CHANNELS),
 	m_modified_channels(ALL_CHANNELS),
 	m_prepare_count(0)
 {
+	m_timer_running[0] = m_timer_running[1] = 0;
+
 	// inform the interface of their engine
 	m_intf.m_engine = this;
 
 	// create the channels
 	for (uint32_t chnum = 0; chnum < CHANNELS; chnum++)
-		m_channel[chnum] = std::make_unique<fm_channel<RegisterType>>(*this, RegisterType::channel_offset(chnum));
+		m_channel[chnum] = new fm_channel<RegisterType>(*this, RegisterType::channel_offset(chnum));
 
 	// create the operators
 	for (uint32_t opnum = 0; opnum < OPERATORS; opnum++)
-		m_operator[opnum] = std::make_unique<fm_operator<RegisterType>>(*this, RegisterType::operator_offset(opnum));
+		m_operator[opnum] = new fm_operator<RegisterType>(*this, RegisterType::operator_offset(opnum));
 
 #if (YMFM_DEBUG_LOG_WAVFILES)
 	for (uint32_t chnum = 0; chnum < CHANNELS; chnum++)
@@ -1229,12 +1230,12 @@
 	write(RegisterType::REG_MODE, 0);
 
 	// reset the channels
-	for (auto &chan : m_channel)
-		chan->reset();
+	for (uint32_t chnum = 0; chnum < CHANNELS; chnum++)
+		m_channel[chnum]->reset();
 
 	// reset the operators
-	for (auto &op : m_operator)
-		op->reset();
+	for (uint32_t opnum = 0; opnum < OPERATORS; opnum++)
+		m_operator[opnum]->reset();
 }
 
 
@@ -1460,7 +1461,7 @@
 		for (uint32_t index = 0; index < 4; index++)
 		{
 			uint32_t opnum = bitfield(map.chan[chnum], 8 * index, 8);
-			m_channel[chnum]->assign(index, (opnum == 0xff) ? nullptr : m_operator[opnum].get());
+			m_channel[chnum]->assign(index, (opnum == 0xff) ? nullptr : m_operator[opnum]);
 		}
 }
 
diff -u -r a/ymfm_opl.cpp b/ymfm_opl.cpp
--- a/ymfm_opl.cpp
+++ b/ymfm_opl.cpp
@@ -430,6 +430,7 @@
 }
 
 
+#if 0 // VOPL3: not built - only the YMF262 (OPL3) is used
 //*********************************************************
 //  OPLL SPECIFICS
 //*********************************************************
@@ -1244,6 +1245,8 @@
 
 
 
+#endif
+
 //*********************************************************
 //  YMF262
 //*********************************************************
@@ -1416,6 +1419,7 @@
 
 
 
+#if 0 // VOPL3: not built - only the YMF262 (OPL3) is used
 //*********************************************************
 //  YMF289B
 //*********************************************************
@@ -2216,5 +2220,6 @@
 
 template class opl_registers_base<4>;
 template class fm_engine_base<opl_registers_base<4>>;
+#endif
 
 }
diff -u -r a/ymfm_opl.h b/ymfm_opl.h
--- a/ymfm_opl.h
+++ b/ymfm_opl.h
@@ -34,9 +34,13 @@
 #pragma once
 
 #include "ymfm.h"
+#if 0 // VOPL3: not built - only the YMF262 (OPL3) is used
 #include "ymfm_adpcm.h"
+#endif
 #include "ymfm_fm.h"
+#if 0 // VOPL3: not built - only the YMF262 (OPL3) is used
 #include "ymfm_pcm.h"
+#endif
 
 namespace ymfm
 {
@@ -273,13 +277,14 @@
 	uint16_t m_waveform[WAVEFORMS][WAVEFORM_LENGTH]; // waveforms
 };
 
-using opl_registers = opl_registers_base<1>;
-using opl2_registers = opl_registers_base<2>;
-using opl3_registers = opl_registers_base<3>;
-using opl4_registers = opl_registers_base<4>;
+typedef opl_registers_base<1> opl_registers;
+typedef opl_registers_base<2> opl2_registers;
+typedef opl_registers_base<3> opl3_registers;
+typedef opl_registers_base<4> opl4_registers;
 
 
 
+#if 0 // VOPL3: not built - only the YMF262 (OPL3) is used
 // ======================> opll_registers
 
 //
@@ -495,13 +500,16 @@
 //  OPL IMPLEMENTATION CLASSES
 //*********************************************************
 
+#endif
+
+#if 0 // VOPL3: not built - only the YMF262 (OPL3) is used
 // ======================> ym3526
 
 class ym3526
 {
 public:
-	using fm_engine = fm_engine_base<opl_registers>;
-	using output_data = fm_engine::output_data;
+	typedef fm_engine_base<opl_registers> fm_engine;
+	typedef fm_engine::output_data output_data;
 	static constexpr uint32_t OUTPUTS = fm_engine::OUTPUTS;
 
 	// constructor
@@ -540,8 +548,8 @@
 class y8950
 {
 public:
-	using fm_engine = fm_engine_base<opl_registers>;
-	using output_data = fm_engine::output_data;
+	typedef fm_engine_base<opl_registers> fm_engine;
+	typedef fm_engine::output_data output_data;
 	static constexpr uint32_t OUTPUTS = fm_engine::OUTPUTS;
 
 	static constexpr uint8_t STATUS_ADPCM_B_PLAYING = 0x01;
@@ -594,8 +602,8 @@
 class ym3812
 {
 public:
-	using fm_engine = fm_engine_base<opl2_registers>;
-	using output_data = fm_engine::output_data;
+	typedef fm_engine_base<opl2_registers> fm_engine;
+	typedef fm_engine::output_data output_data;
 	static constexpr uint32_t OUTPUTS = fm_engine::OUTPUTS;
 
 	// constructor
@@ -635,13 +643,15 @@
 //  OPL3 IMPLEMENTATION CLASSES
 //*********************************************************
 
+#endif
+
 // ======================> ymf262
 
 class ymf262
 {
 public:
-	using fm_engine = fm_engine_base<opl3_registers>;
-	using output_data = fm_engine::output_data;
+	typedef fm_engine_base<opl3_registers> fm_engine;
+	typedef fm_engine::output_data output_data;
 	static constexpr uint32_t OUTPUTS = fm_engine::OUTPUTS;
 
 	// constructor
@@ -677,6 +687,7 @@
 };
 
 
+#if 0 // VOPL3: not built - only the YMF262 (OPL3) is used
 // ======================> ymf289b
 
 class ymf289b
@@ -684,8 +695,8 @@
 	static constexpr uint8_t STATUS_BUSY_FLAGS = 0x05;
 
 public:
-	using fm_engine = fm_engine_base<opl3_registers>;
-	using output_data = fm_engine::output_data;
+	typedef fm_engine_base<opl3_registers> fm_engine;
+	typedef fm_engine::output_data output_data;
 	static constexpr uint32_t OUTPUTS = 2;
 
 	// constructor
@@ -746,9 +757,9 @@
 	static constexpr uint32_t FM_EXTRA_SAMPLE_STEP = 192 - FM_EXTRA_SAMPLE_THRESH;
 
 public:
-	using fm_engine = fm_engine_base<opl4_registers>;
+	typedef fm_engine_base<opl4_registers> fm_engine;
 	static constexpr uint32_t OUTPUTS = 6;
-	using output_data = ymfm_output<OUTPUTS>;
+	typedef ymfm_output<OUTPUTS> output_data;
 
 	static constexpr uint8_t STATUS_BUSY = 0x01;
 	static constexpr uint8_t STATUS_LD = 0x02;
@@ -803,8 +814,8 @@
 class opll_base
 {
 public:
-	using fm_engine = fm_engine_base<opll_registers>;
-	using output_data = fm_engine::output_data;
+	typedef fm_engine_base<opll_registers> fm_engine;
+	typedef fm_engine::output_data output_data;
 	static constexpr uint32_t OUTPUTS = fm_engine::OUTPUTS;
 
 	// constructor
@@ -897,6 +908,8 @@
 	static uint8_t const s_default_instruments[];
 };
 
+#endif
+
 }
 
 #endif // YMFM_OPL_H
